ceno_zkvm/src/instructions/riscv/div.rs

```diff
@@ -1,124 +1,348 @@
 
-pub struct RemuOp;
-impl RIVInstruction for RemuOp {
-    const INST_KIND: InsnKind = InsnKind::REMU;
+pub struct DivOp;
+impl RIVInstruction for DivOp {
+    const INST_KIND: InsnKind = InsnKind::DIV;
 }
-pub type RemuDummy<E> = DummyInstruction<E, RemuOp>; // TODO: implement RemuInstruction.
+pub type DivInstruction<E> = ArithInstruction<E, DivOp>;
 
 impl<E: ExtensionField, I: RIVInstruction> Instruction<E> for ArithInstruction<E, I> {
-    type InstructionConfig = ArithConfig<E>;
+    type InstructionConfig = DivRemConfig<E>;
 
     fn name() -> String {
         format!("{:?}", I::INST_KIND)
     }
 
     fn construct_circuit(cb: &mut CircuitBuilder<E>) -> Result<Self::InstructionConfig, ZKVMError> {
-        // outcome = dividend / divisor + remainder => dividend = divisor * outcome + r
-        let mut divisor = UInt::new_unchecked(|| "divisor", cb)?;
-        let mut outcome = UInt::new(|| "outcome", cb)?;
-        let r = UInt::new(|| "remainder", cb)?;
-        let (dividend, inter_mul_value) =
-            divisor.mul_add(|| "divisor * outcome + r", cb, &mut outcome, &r, true)?;
-
-        // div by zero check
-        let is_zero =
-            IsZeroConfig::construct_circuit(cb, || "divisor_zero_check", divisor.value())?;
-        let outcome_value = outcome.value();
+        // The soundness analysis for these constraints is only valid for
+        // 32-bit registers represented over the Goldilocks field, so verify
+        // these parameters
+        assert_eq!(UInt::<E>::TOTAL_BITS, u32::BITS as usize);
+        assert_eq!(E::BaseField::MODULUS_U64, goldilocks::MODULUS);
+
+        let dividend = UInt::new_unchecked(|| "dividend", cb)?; // 32-bit value from rs1
+        let divisor = UInt::new_unchecked(|| "divisor", cb)?; // 32-bit value from rs2
+        let quotient = UInt::new(|| "quotient", cb)?;
+        let remainder = UInt::new(|| "remainder", cb)?;
+
+        // `rem_e` and `div_e` are expressions representing the remainder and
+        // divisor from the signed or unsigned division operation, with signs
+        // renormalized to be nonnegative, so that correct values must satisfy
+        // either `0 <= rem_e < div_e` or `div_e == 0`.  The `rem_e` value
+        // should be constrained to be nonnegative before being returned from
+        // this block, while the checks `rem_e < div_e` or `div_e == 0` are
+        // done later.
+        let (internal_config, rem_e, div_e) = match I::INST_KIND {
+            InsnKind::DIVU | InsnKind::REMU => {
+                cb.require_equal(
+                    || "unsigned_division_relation",
+                    dividend.value(),
+                    divisor.value() * quotient.value() + remainder.value(),
+                )?;
+
+                (InternalDivRem::Unsigned, remainder.value(), divisor.value())
+            }
+
+            InsnKind::DIV | InsnKind::REM => {
+                let dividend_signed: Signed<E> =
```

Comment:
Great, I definitely tend to overuse type annotations in some contexts -- thanks for pointing it out.

Reply:
It's not too much of a problem here actually, because the type is short and doesn't really contribute to the verbosity of the code.

(And it's a bit of a judgement call to decide when extra type annotations help with understanding, and when they hinder with verbosity.)